Mathlib/Order/Directed.lean

```diff
@@ -321,3 +325,57 @@ instance (priority := 100) OrderTop.to_isDirected_le [LE α] [OrderTop α] : IsD
 -- see Note [lower instance priority]
 instance (priority := 100) OrderBot.to_isDirected_ge [LE α] [OrderBot α] : IsDirected α (· ≥ ·) :=
   ⟨fun _ _ => ⟨⊥, bot_le _, bot_le _⟩⟩
+
+namespace DirectedOn
+
+section Pi
+
+variable {ι : Type*} {α : ι → Type*}
+
+variable (r : (i : ι) →  α i →  α i → Prop)
+
+lemma proj {d : Set (Π i, α i)} (hd : DirectedOn (fun x y => ∀ i, r i (x i)  (y i)) d) (i : ι) :
+    DirectedOn (r i) ((fun a => a i) '' d) :=
+  DirectedOn.mono_comp (fun _ _ h => h) (mono hd fun ⦃_ _⦄ h ↦ h i)
+
+lemma pi {d : (i : ι) → Set (α i)} (hd : ∀ (i : ι), DirectedOn (r i) (d i)) :
+    DirectedOn (fun x y => ∀ i, r i (x i) (y i)) (Set.pi  Set.univ d) := by
+  intro a ha b hb
+  choose f hf using fun i => hd i (a i) (ha i trivial) (b i) (hb i trivial)
+  simpa [Pi.le_def, ← forall_and] using ⟨f, fun i =>
+    ⟨by simp_all only [Set.mem_pi, Set.mem_univ, forall_const],
+       by constructor <;> (simp_all only [Set.mem_pi, Set.mem_univ, forall_const])⟩⟩
+
+end Pi
+
+section Prod
+
+variable {r₂ : β → β → Prop}
+
+/-- Local notation for a relation -/
+local infixl:50 " ≼₁ " => r
+/-- Local notation for a relation -/
+local infixl:50 " ≼₂ " => r₂
+
+lemma fst {d : Set (α × β)} (hd : DirectedOn (fun p q ↦ p.1 ≼₁ q.1 ∧ p.2 ≼₂ q.2) d) :
+    DirectedOn (· ≼₁ ·) (Prod.fst '' d) := by
+  intro a ⟨p,hp⟩ b ⟨q,hq⟩
+  obtain ⟨r,hr⟩ := hd p hp.1 q hq.1
+  aesop
+
+lemma snd {d : Set (α × β)} (hd : DirectedOn (fun p q ↦ p.1 ≼₁ q.1 ∧ p.2 ≼₂ q.2) d) :
+    DirectedOn (· ≼₂ ·) (Prod.snd '' d) := by
+  intro a ⟨p,hp⟩ b ⟨q,hq⟩
+  obtain ⟨r,hr⟩ := hd p hp.1 q hq.1
+  aesop
+
+lemma prod {d₁ : Set α} {d₂ : Set β} (h₁ : DirectedOn (· ≼₁ ·) d₁) (h₂ : DirectedOn (· ≼₂ ·) d₂) :
+    DirectedOn (fun p q ↦ p.1 ≼₁ q.1 ∧ p.2 ≼₂ q.2) (d₁ ×ˢ d₂) := fun _ hpd _ hqd => by
+  obtain ⟨r₁,hr₁⟩ := h₁ _ hpd.1 _ hqd.1
+  obtain ⟨r₂,hr₂⟩ := h₂ _ hpd.2 _ hqd.2
+  use (r₁, r₂)
+  aesop
```

Comment:
As much as I like `aesop`, I think using it here is a bit overkill.
```suggestion
  obtain ⟨r₁, hdr₁, hpr₁, hqr₁⟩ := h₁ _ hpd.1 _ hqd.1
  obtain ⟨r₂, hdr₂, hpr₂, hqr₂⟩ := h₂ _ hpd.2 _ hqd.2
  exact ⟨⟨r₁, r₂⟩, ⟨hdr₁, hdr₂⟩, ⟨hpr₁, hpr₂⟩, ⟨hqr₁, hqr₂⟩⟩
```

But maybe I'm too infected by angle-brackets, so feel free to ignore if you disagree.